### PR TITLE
enable `consider-using-enumerate` rule and fix pylint errors

### PR DESCRIPTION
**Pull request by ayushthe1**
Signed-off-by: ayush_gitk <ayushsharmaa101@gmail.com>

## Related Issues/PRs

 Solves a part of #7205

## What changes are proposed in this pull request?

enabled the rule  [`consider-using-enumerate`](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/consider-using-enumerate.html) and fixed the pylint errors.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 1]
```diff
@@ -662,8 +662,8 @@ def test_create_run_appends_to_artifact_uri_path_correctly(self):
         # Patch `is_local_uri` to prevent the SqlAlchemy store from attempting to create local
         # filesystem directories for file URI and POSIX path test cases
         with mock.patch("mlflow.store.tracking.sqlalchemy_store.is_local_uri", return_value=False):
-            for i in range(len(cases)):
-                artifact_root_uri, expected_artifact_uri_format = cases[i]
+            for _, case in enumerate(cases):
+                artifact_root_uri, expected_artifact_uri_format = case
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:666]:
```suggestion
            for artifact_root_uri, expected_artifact_uri_format in cases:
```

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 2]
```diff
@@ -600,8 +600,8 @@ def test_create_experiment_appends_to_artifact_uri_path_correctly(self):
...
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:604]: Same here https://github.com/mlflow/mlflow/pull/7306/files#r1019184930

File: tests/store/tracking/test_abstract_store.py [hunk 3]
```diff
@@ -89,8 +89,8 @@ def test_search_runs():
     with mock.patch.object(AbstractStoreTestImpl, "_search_runs", return_value=(runs, token)):
         store = AbstractStoreTestImpl()
         result = store.search_runs([experiment_id], None, view_type)
-        for i in range(len(result)):
-            assert result[i] == runs[i]
+        for i, res in enumerate(result):
+            assert res == runs[i]
```
Comment (harupy) [inline tests/store/tracking/test_abstract_store.py:93]:
```suggestion
        assert list(result) == runs
```

Does this work?

Comment (ayushthe1) [reply tests/store/tracking/test_abstract_store.py]: better as it reduces complexity . :+1:

**Comment by ayushthe1** [comment]
Thanks @harupy , fixed the suggestions.

**Comment by ayushthe1** [comment]
image
<br>
hey @harupy ,can i use something like list-comprehension to solve this ?
```
actual_buckets_low_value = [x.low_value for x in actual.buckets]
expected_buckets_low_value = [x.low_value for x in expected.buckets]
assert actual_buckets_low_value == pytest.approx(expected_buckets_low_value ,rel=1e-2)
```

File: tests/store/tracking/test_abstract_store.py [hunk 4]
```diff
@@ -89,8 +89,7 @@ def test_search_runs():
     with mock.patch.object(AbstractStoreTestImpl, "_search_runs", return_value=(runs, token)):
         store = AbstractStoreTestImpl()
         result = store.search_runs([experiment_id], None, view_type)
-        for i in range(len(result)):
-            assert result[i] == runs[i]
+        assert list[result] == runs
```
Comment (BenWilson2) [inline tests/store/tracking/test_abstract_store.py:92]:
```suggestion
        assert list(result) == runs
```
typo :)

**Comment by BenWilson2** [comment]
> image hey @harupy ,can i use something like list-comprehension to solve this ?
> 
> ```
> actual_buckets_low_value = [x.low_value for x in actual.buckets]
> expected_buckets_low_value = [x.low_value for x in expected.buckets]
> assert actual_buckets_low_value == pytest.approx(expected_buckets_low_value ,rel=1e-2)
> ```

A straight enumeration with index value will likely be simpler and easier to read :) 

``` python
for i, exp_bucket in enumerate(expected.buckets):
  act_bucket = actual.buckets[i]
  assert exp_bucket.low_value == pytest.approx(act_bucket.low_value, rel=1e-2)
```

File: tests/recipes/cards/test_histogram_generator.py [hunk 5]
```diff
@@ -22,14 +22,11 @@ def assert_histogram(
...
-            assert actual_bucket.low_value == pytest.approx(expected_bucket.low_value, rel=1e-2)
-            assert actual_bucket.high_value == pytest.approx(expected_bucket.high_value, rel=1e-2)
-            assert actual_bucket.sample_count == pytest.approx(
-                expected_bucket.sample_count, rel=1e-2
-            )
+        for i, exp_bucket in enumerate(expected.buckets):
+            act_bucket = actual.buckets[i]
+            assert act_bucket.low_value == pytest.approx(exp_bucket.low_value, rel=1e-2)
+            assert act_bucket.high_value == pytest.approx(exp_bucket.high_value, rel=1e-2)
+            assert act_bucket.sample_count == pytest.approx(exp_bucket.sample_count, rel=1e-2)
```
Comment (harupy) [inline tests/recipes/cards/test_histogram_generator.py:29]:
```suggestion
        for acut_bucket, exp_bucket in zip(actual.buckets, expected.buckets):
            assert act_bucket.low_value == pytest.approx(exp_bucket.low_value, rel=1e-2)
            assert act_bucket.high_value == pytest.approx(exp_bucket.high_value, rel=1e-2)
            assert act_bucket.sample_count == pytest.approx(exp_bucket.sample_count, rel=1e-2)
```

I'd use zip here.

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by ayushthe1** [comment]
> LGTM!

thanks .
hey @harupy  ,currently i know about tensorflow and sklearn .Is there anything related to it i can contribute ?

**Comment by harupy** [comment]
> hey @harupy ,currently i know about tensorflow and sklearn .Is there anything related to it i can contribute ?

Let me find something that you can contribute to.

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/actions/runs/3437135162/jobs/5731524214 might help find contribution ideas.

```
👇
  /opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/sklearn/cluster/_kmeans.py:867: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
    warnings.warn(

tests/sklearn/test_sklearn_autolog.py: 34 warnings
  /opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
    warnings.warn(

👇
tests/sklearn/test_sklearn_autolog.py::test_basic_post_training_metric_autologging
  /opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/sklearn/metrics/_scorer.py:780: FutureWarning: sklearn.metrics.SCORERS is deprecated and will be removed in v1.3. Please use sklearn.metrics.get_scorer_names to get a list of available scorers and sklearn.metrics.get_metric to get scorer.
```

We have some deprecation warnings that should be fixed.

**Pull request merged**